Updated approach to counting the number of rows for replicated tables at DML

red1452 opened the pull request:
Update approach to counting the number of rows for replicated tables at DML

Replicated tables are the same at all segments. Modifying replicated tables lead
to do the same processing at each segment. All segments calculate the count of
modified (updated, deleted or inserted) rows. The count on all segments will be
the same. It led to situation: QD got the same counts of modified rows from each
segments, summarized it, then divided it on number of segments. But it had to do
not in all cases. One such case: select from a function which contains DML on
replicated table. Such query led to an error: "consistency check on SPI tuple
count failed".

Also, the ExecutePlan function allows to limit the number of processing tuples,
but for DML operations the number of tuples is acquired from the sum of such
numbers from segments. It led to the same error, because QD processed the number
of tuples, which is less than all segments sent.

This patch modifies approach to counting the number of rows for replicated
tables at DML operations. Only one segment will calculate the count of modified
rows if the table is replicated. Other segments will send zero. It was done by
disabling the canSetTag flag for all segments except one. Dividing on number of
segments is no longer needed.

Also, for SPI calculating the number of processed tuples was added for cases,
when it is used at the _SPI_checktuples function. The behavior of _SPI_pquery
was changed to the behavior at PostgreSQL.

NOTE: the canSetTag flag enables saving the lastTid variable for the INSERT
queries. But it is useless, because it is used at only one place: at the
currtid_byreloid function, which has been never used. It was removed by
PostgreSQL 14:
https://github.com/postgres/postgres/commit/7b94e999606e2e2e10d68d544d49fc5a5d5785ac

red1452 committed:
Fix error at select from replicated table in function

For replicated tables there is adjusting, which divides count of processed
tuples on number of segments at QD at the end of query execution for DML
operations. Such adjusting was applied, when we do select from function, which
contains DML operation with replicated table. It was incorrect behavior, because
the main operation is SELECT.

This patch add check, which allows not to do adjusting if current query is not
DML.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/72349

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -3056,7 +3056,7 @@ ExecPostprocessPlan(EState *estate)
  */
 void
-ExecEndPlan(PlanState *planstate, EState *estate)
+ExecEndPlan(PlanState *planstate, EState *estate, CmdType operation)
```
how about to change it to
```suggestion
ExecEndPlan(QueryDesc *queryDesc)
```
because QueryDesc contains PlanState, EState and CmdType?

red1452 replied:
`Planstate` has pointer to `estate`, but does not have field `operation`. And for execution `ExecEndPlan`, pointer to `estate` is get from `queryDesc`. Do you sure, that these are the same pointers at all time?

RekGRpth replied:
> `Planstate` has pointer to `estate`, but does not have field `operation`. And for execution `ExecEndPlan`, pointer to `estate` is get from `queryDesc`. Do you sure, that these are the same pointers at all time?

I corrected suggestion, look more carefully

red1452 replied:
done.

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1479689

red1452 committed:
reduce count of args at ExecEndPlan

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/72412

RekGRpth commented on src/backend/executor/execMain.c:
```diff
@@ -3056,8 +3056,10 @@ ExecPostprocessPlan(EState *estate)
-ExecEndPlan(PlanState *planstate, EState *estate)
+ExecEndPlan(QueryDesc *queryDesc)
 {
+	PlanState *planstate = queryDesc->planstate;
+	EState *estate = queryDesc->estate;
```
may be add some tabs/spaces before * like in indent?

RekGRpth commented:
how about to simplify test
```sql
DO $$
DECLARE
    res integer;
BEGIN
    CREATE TEMPORARY TABLE t_test(a integer) DISTRIBUTED REPLICATED;

    WITH cte AS (
        INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a
    ) SELECT sum(a) FROM cte INTO res;

    RAISE NOTICE '%', res;

    DROP TABLE t_test;
END;
$$;
```
(taken from linked ticked)

RekGRpth commented:
on the other hand, patch does not solve this problem
```sql
DO $$
DECLARE
    res integer;
BEGIN
    CREATE TEMPORARY TABLE t_test(a integer) DISTRIBUTED REPLICATED;

    INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a INTO res;

    RAISE NOTICE '%', res;

    DROP TABLE t_test;
END;
$$;
ERROR:  consistency check on SPI tuple count failed (spi.c:2680)
CONTEXT:  SQL statement "INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a"
PL/pgSQL function inline_code_block line 7 at SQL statement
```

red1452 commented:
> on the other hand, patch does not solve this problem
> 
> ```sql
> DO $$
> DECLARE
>     res integer;
> BEGIN
>     CREATE TEMPORARY TABLE t_test(a integer) DISTRIBUTED REPLICATED;
> 
>     INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a INTO res;
> 
>     RAISE NOTICE '%', res;
> 
>     DROP TABLE t_test;
> END;
> $$;
> ERROR:  consistency check on SPI tuple count failed (spi.c:2680)
> CONTEXT:  SQL statement "INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a"
> PL/pgSQL function inline_code_block line 7 at SQL statement
> ```



> on the other hand, patch does not solve this problem
> 
> ```sql
> DO $$
> DECLARE
>     res integer;
> BEGIN
>     CREATE TEMPORARY TABLE t_test(a integer) DISTRIBUTED REPLICATED;
> 
>     INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a INTO res;
> 
>     RAISE NOTICE '%', res;
> 
>     DROP TABLE t_test;
> END;
> $$;
> ERROR:  consistency check on SPI tuple count failed (spi.c:2680)
> CONTEXT:  SQL statement "INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a"
> PL/pgSQL function inline_code_block line 7 at SQL statement
> ```

Your query is not correct. If you want to write result to a variable. You must be sure, that your `INSERT` returns only one row. But at your query `INSERT` returns 3 rows. I agree, that message is not informative, but after fix your query will fail, but with the other message. If little rewrite your query - it works:
```sql
postgres=# DO $$
DECLARE
    res integer;
BEGIN
    CREATE TEMPORARY TABLE t_test(a integer) DISTRIBUTED REPLICATED;

    INSERT INTO t_test(a) VALUES (1) RETURNING a INTO res;

    RAISE NOTICE '%', res;

    DROP TABLE t_test;
END;
$$;
NOTICE:  1
DO
```
```

red1452 committed:
Updated approach to counting the number of rows for replicated tables at DML

Replicated tables are the same at all segments. Modifying replicated tables lead
to do the same processing at each segment. All segments calculate the count of
modified (updated, deleted or inserted) rows. The count on all segments will be
the same. It led to situation: QD got the same count of modified rows from each
segments, summarize it, then divide it on number of segments. But it had to do
not in all cases. One such case: select from a function which contains DML on
replicated table. Such query led to an error: "consistency check on SPI tuple
count failed".

Also, the ExecutePlan function allows to limit the number of processing tuples,
but for DML operations the number of tuples is get from summ of such numbers
from segments. It led to the same error, because QD processed the number of
tuples, which is less than all segments sent.

This patch modifies approach to counting the number of rows for replicated
tables at DML operations. Only one segment will calculate the count of modified
rows if the table is replicated. Other segments will send zero. Dividing on
number of segments is no longer needed. Also, saving the number of current
processed tuples to the all processed tuples was added if processing tuples was
broken by the limit.

red1452 committed:
remove extra changes

red1452 committed:
add one more test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/72749

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1500165

RekGRpth commented on src/test/regress/sql/bfv_dml_rpt.sql:
```diff
@@ -0,0 +1,76 @@
+INSERT INTO dstr_repl_tst_stg VALUES (3, 'III'), (4, 'IV'), (5, 'V'), (6, 'VI'), (7, 'VII');
+
+-- create merger function
+CREATE OR REPLACE FUNCTION MergeToEcom(OUT u INT, OUT i INT) RETURNS record as
+$$ BEGIN
+
+    WITH updated AS (
+        UPDATE dstr_repl_tst d
+        SET DivisionName = s.DivisionName
+        FROM dstr_repl_tst_stg s
+        WHERE d.DivisionID = s.DivisionID
+        returning 1)
+    SELECT count(*) into u from updated;
+    i := 0;
+
+END $$ language plpgsql;
```
how about to simplify test
```sql
DO $$
DECLARE
    res integer;
BEGIN
    CREATE TEMPORARY TABLE t_test(a integer) DISTRIBUTED REPLICATED;

    WITH cte AS (
        INSERT INTO t_test(a) VALUES (1), (2), (3) RETURNING a
    ) SELECT sum(a) FROM cte INTO res;

    RAISE NOTICE '%', res;

    DROP TABLE t_test;
END;
$$;
```
(taken from linked ticket)

RekGRpth commented on src/backend/executor/execUtils.c:
```diff
@@ -2163,8 +2163,14 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
+			/*
+			 * Do not modify es_processed if it is not zero. Because
+			 * es_processed save info about tuples, which were really processed,
+			 * and limit was used (not to process tuples more than limit).
```
Is there a check for this in tests?

red1452 replied:
Yes, it checks by block with comment: `Check, that INSERT produce the error with correct explanation`

RekGRpth commented:
I think it is not good:
```sql
CREATE TABLE t(a int) DISTRIBUTED REPLICATED;
INSERT INTO t(a) VALUES (1), (2), (3);
```
```sh
PGOPTIONS='-c gp_session_role=utility' psql -p 6002
```
```sql
select * from t where a = 2;
 a 
---
 2
(1 row)

update t set a = 1 where a = 2;
UPDATE 0
select * from t where a = 2;
 a 
---
(0 rows)
```

red1452 committed:
fix calculating es_porcessed at utility, optimize test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/72936

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -632,7 +632,17 @@ ExecInsert(TupleTableSlot *parentslot,
+		 * It allows not to adjust es_processed at QD after all executors send
+		 * the same value of es_processed.
+		 */
+		struct GpPolicy * cdbpolicy = resultRelInfo->ri_RelationDesc->rd_cdbpolicy;
+		if (Gp_role != GP_ROLE_EXECUTE || !GpPolicyIsReplicated(cdbpolicy) ||
+		    GpIdentity.segindex == (gp_session_id % cdbpolicy->numsegments))
```
avoid local var for policy, according to https://github.com/arenadata/gpdb/commit/0bfc72517d7a28d5a4143ba026f57e5f35caf29e

red1452 committed:
remove local variable

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-5691

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1516953

red1452 committed:
optimization

red1452 committed:
Merge remote-tracking branch 'origin/adb-6.x-dev' into ADBDEV-5691

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/73361

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1534913

whitehawk commented:
Latest failures on CI with ORCA look related to the patch, as these faulires are related to tuple counting on replicated tables:
```diff
--- \/home\/gpadmin\/gpdb_src\/src\/test\/isolation2\/expected\/parallel_retrieve_cursor\/explain\.out	2024-06-18 13:15:36.654746453 +0000
+++ \/home\/gpadmin\/gpdb_src\/src\/test\/isolation2\/results\/parallel_retrieve_cursor\/explain\.out	2024-06-18 13:15:36.666746743 +0000
@@ -16,7 +16,7 @@
 CREATE TABLE rt1 (a INT) DISTRIBUTED REPLICATED;
 CREATE
 insert into rt1 select generate_series(1,100);
-INSERT 100
+INSERT 300
 
 -- PARALLEL RETRIEVE CURSOR with other options (WITH HOLD/SCROLL) is not supported
 EXPLAIN (COSTS false) DECLARE c1 PARALLEL RETRIEVE CURSOR WITHOUT HOLD FOR SELECT * FROM t1;

======================================================================

--- \/home\/gpadmin\/gpdb_src\/src\/test\/isolation2\/expected\/parallel_retrieve_cursor\/replicated_table\.out	2024-06-18 13:15:47.983020071 +0000
+++ \/home\/gpadmin\/gpdb_src\/src\/test\/isolation2\/results\/parallel_retrieve_cursor\/replicated_table\.out	2024-06-18 13:15:47.999020457 +0000
@@ -6,7 +6,7 @@
 CREATE TABLE rt1 (a INT) DISTRIBUTED REPLICATED;
 CREATE
 insert into rt1 select generate_series(1,100);
-INSERT 100
+INSERT 300
 
 --------- Test1: Basic test for PARALLEL RETRIEVE CURSOR on replicated table
 

======================================================================
```

whitehawk commented on src/test/regress/greenplum_schedule:
```diff
@@ -189,7 +189,7 @@ test: bb_memory_quota memconsumption
 test: tuple_serialization
 # Tests for replicated table
-test: rpt rpt_joins rpt_tpch rpt_returning
+test: rpt rpt_joins rpt_tpch rpt_returning bfv_dml_rpt
```
why do we need a separate test file? Can't we add this test to some existing file (for ex. 'bfv_dml.sql') and have less files touched and less entities?

red1452 replied:
I think, it is better to make separate tests, because big files is worse to understand than small.

red1452 committed:
make spi closier to postgres

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/73663

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1554235

whitehawk commented:
In the description:

> " It led to situation: QD got the same counts of modified rows from each
segments, summarize it, then divide it on number of segments. "

Let's keep the past tense consistent. "summarize" -> "summarized" and "divide" -> "divided".

> "but for DML operations the number of tuples is get from summ of such numbers"

Passive voice requires past participle.

"is get from summ " -> "is got from the sum " (or "is acquired from the sum ").

whitehawk commented on src/backend/executor/execMain.c:
```diff
@@ -3202,6 +3199,8 @@ ExecutePlan(EState *estate,
 		/* Reset the per-output-tuple exprcontext */
 		ResetPerTupleExprContext(estate);
 
+		saved_es_processed = estate->es_processed;
```
why not:
```diff
diff --git a/src/backend/executor/execMain.c b/src/backend/executor/execMain.c
index a6a7f12b2c..e4d2bac37a 100644
--- a/src/backend/executor/execMain.c
+++ b/src/backend/executor/execMain.c
@@ -3148,7 +3148,6 @@ ExecutePlan(EState *estate,
        TupleTableSlot *slot;
        long            current_tuple_count;
        ListCell *lc;
-       uint64          saved_es_processed;
 
        /*
         * For holdable cursor, the plan is executed without rewinding on gpdb. We
@@ -3196,11 +3195,11 @@ ExecutePlan(EState *estate,
         */
        for (;;)
        {
+               uint64 saved_es_processed = estate->es_processed;
+
                /* Reset the per-output-tuple exprcontext */
                ResetPerTupleExprContext(estate);
 
-               saved_es_processed = estate->es_processed;
-
                /*
                 * Execute the plan and obtain a tuple
                 */
```
?

whitehawk commented on src/backend/executor/nodeDML.c:
```diff
@@ -327,6 +322,20 @@ ExecInitDML(DML *node, EState *estate, int eflags)
+	if (Gp_role == GP_ROLE_EXECUTE)
+	{
+		if (GpPolicyIsReplicated(resultRelInfo->ri_RelationDesc->rd_cdbpolicy) &&
+			GpIdentity.segindex != (gp_session_id % resultRelInfo->ri_RelationDesc->rd_cdbpolicy->numsegments))
```
Why do we always re-calculate segindex that updates `es_processed`? Can't we just set it to segment index 0?

red1452 replied:
It was taken from the `Explicit gather motion` node logic: https://github.com/arenadata/gpdb/blob/8f17c3762cd88d448b5c6ec8e24114cf97d5a04c/src/backend/executor/nodeMotion.c#L330

whitehawk commented:
Just FYI - interesting thing (reproduced without your delta):

```
postgres=# create table t (a int) distributed replicated;
CREATE TABLE
postgres=# insert into t values (1), (2), (3);
INSERT 0 3
postgres=# with cte as (insert into t values (-1), (-2), (-3) returning *) select * from cte union select * from t;
ERROR:  unexpected arguments to set operation (cdbsetop.c:229)
```

Created ADBDEV-5846 for this problem.

red1452 committed:
small code optimiztion

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-5691

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/73788

dkovalev1 commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -2394,6 +2394,21 @@ ExecInitModifyTable(ModifyTable *node, EState *estate, int eflags)
+	 */
+	if (Gp_role == GP_ROLE_EXECUTE)
+	{
+		if (GpPolicyIsReplicated(mtstate->resultRelInfo->ri_RelationDesc->rd_cdbpolicy) &&
```
It would be cleaner if extract mtstate->resultRelInfo->ri_RelationDesc->rd_cdbpolicy to temp variable.

RekGRpth replied:
https://github.com/arenadata/gpdb/pull/969#discussion_r1637529656

red1452 committed:
small fix

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1617567

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1617566

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-5691

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/74744
